Bucket quota should be treated and stored as BigInts

williamlardier opened the pull request:
Stored and treated as BigInt in the class. Accepted as string/number/Long (from a serialized object, or from an object), and outputted as a string for storage on the DB

Seragonia committed:
Configure project to accept latest BigInt-related notation

Issue: ARSN-457

Seragonia committed:
Switch quotas to BigInt

The previous objects will be handled because we always convert
incoming data (inDeSerialize and fromObj) to BigInts, which is
working well if the value is a string, a number, or any mongodb-
specific value.

Issue: ARSN-457

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-457 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.1.146`

* `8.2.0`


Please check the `Fix Version/s` of ARSN-457, or the target
branch of this pull request.

Seragonia committed:
Fix ubuntu runner os version to 22

Libcrypto was removed, plus other changes, not yet supported
by Arsenal.

Issue: ARSN-457

francoisferrand commented on lib/auth/Vault.ts:
```diff
@@ -18,7 +18,10 @@ function vaultSignatureCb(
                 userInfo: AuthInfoType,
                 authorizationResults: { [key: string]: any },
-                accountQuota: number,
+                accountQuota: {
```
API format changed?
this is not just passing a string instead of number (or ideally `number|string` for backwards compatibility), but a real compatibility break...

williamlardier replied:
No, the type was wrong from the beginning, it was already an object with number, now it's a string.

williamlardier replied:
It would have created errors if we had TS in cloudserver.

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -188,7 +188,7 @@ export default class BucketInfo {
         tags?: Array<BucketTag> | [],
         capabilities?: Capabilities,
-        quotaMax?: number | 0,
+        quotaMax?: bigint,
```
should we not be more leniant, to avoid issues: e.g. allow the api to take `number | bigint | 0`, and initalize the field with `this._quotaMax = BigInt(quotaMax || 0n)` ?
or is there a performance benefit to doing this?

(and same for setQuota api)

williamlardier replied:
The API accepts strings only since the beginning. Here, this is just how we parse/handle the value internally, it must be bigint here because we control how we create the BucketInfo class, no need to keep debt here.

Seragonia committed:
Support SOSAPI metrics as large numbers

- Long in MongoDB
- String when retrieved / serialized
- BigInt locally

Issue: ARSN-457

Seragonia committed:
Consistent types for Auth APIs

Issue: ARSN-457

Seragonia committed:
Be lenient on the input BucketInfo type for bucket quota

Issue: ARSN-457

Seragonia committed:
Accept bigint and numbers for the veeamsosapi capability

Issue: ARSN-457

Seragonia committed:
Do not use the toJSON

- We were passing a test on serialize() thank to the global
  toJSON defined in the test, yet during execution time, we
  would not be able to serialize. We cannot use toJSON globally
  in this repo, because it would affect all projects importing
  it, and eventually lead to unexpected side effects.
- The manual conversion to string must be carefully done: in
  nodejs, converting to json a bigint as a subfield of an object
  would change the base object in the class, and eventually we
  end up with "strings" instead of bigint, without typescript
  knowing about it.

Issue: ARSN-457

Seragonia committed:
Migrate MongoClientInterface to typescript

- Allows us to have a better confidence over the mongo methods
- Helped fix numerous bugs:
 - Disconnection from mongodb was not detected properly.
 - Getting collection might fail without error handling.
 - Bucket Version in-memory cache was never properly set.
 - Arsenal error handling was lacking checks to ensure the
   error is of the right type.
- We now better control the types in mongo, so we can ensure
  we properly handle specific types like Long affecting both
  the bucket quota and the capabilities.

Issue: ARSN-457

Seragonia committed:
Fix BucketInfo class

- S3Utils is testing the fact that the capabilities might be
  set to -1, as the default value.
- Have a default when converting a variable to a bigint.

Issue: ARSN-457

williamlardier commented on lib/algos/list/skip.js:
```diff
@@ -10,10 +10,10 @@ const MAX_STREAK_LENGTH = 100;
-     *                                    equal) used by the client code
+     * @param {Object} params                - skip parameters
+     * @param {Object} params.extension      - delimiter extension used (required)
+     * @param {String | string[]} params.gte - current range gte (greater than or
```
This change is related to the fact that in MongoClientInterface we actually pass an array here: the function type was not properly defined here, leading to TS error. Having an array here is fine and expected.

williamlardier commented on lib/auth/backends/in_memory/Backend.ts:
This file is not actually used (and not in production as well), but we want to have a similar behavior here as in vaultclient, because both are under the same abstraction layer.

Seragonia committed:
Improve types and linting

Issue: ARSN-457

francoisferrand commented on lib/auth/backends/in_memory/Backend.ts:
```diff
@@ -61,7 +70,15 @@ class InMemoryBackend extends BaseBackend {
+                account: entity.canonicalID,
+                // account quota is not supported in this backend
+                accountQuota: 0,
+            }
```
nit: would it not be simpler to just add a `cannonicalId` or `entity` parameter to `formatResponse`, and let that function handle the formatting of nested types and filling of missed fields just once?

```
function _formatResponse(userInfoToSend: AuthInfoType, account: string): { message: { body: AuthV4Results } } {
    return {
        message: {
            body: {
                userInfo: userInfoToSend,
                authorizationResults: [],
                accountQuota: {
                    account,
                    // account quota is not supported in this backend
                    accountQuota: 0,
                },
            },
        },
    };
}
```

```suggestion
const vaultReturnObject = this.formatResponse(userInfoToSent, entity.canonicalID);
```

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -60,22 +63,83 @@ export type VeeamSOSApi = {
 
+// Bigints are exported as strings
+export type CapacityInfoExported = Omit<VeeamSOSApiSchema, 'CapacityInfo'> & {
+    CapacityInfo?: {
```
what do you mean "exported" ?
it creates confusion with javascript export, and does help to understand where it would be in this stringified form: in mongo? when serialized to json? in "network" messages?...

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -38,7 +38,10 @@ export type VersioningConfiguration = {
-export type VeeamSOSApi = {
+// The input VeeamSOSAPI type, where the capcity-related
+// field accept numbers, but will be treated as bigints
+// internally.
```
this add a line to explain where this 'schema' applies : is it to the (javascript) API ? to the data we get/put in mongo ? ...

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -60,22 +63,83 @@ export type VeeamSOSApi = {
+// where the capacity-related fields can be numbers.
+export type CapabilitiesInput = {
+    VeeamSOSApi?: VeeamSOSApiSchema,
+};
```
do we need a separate type here, or could we inline it inside `ParsableBucketMetadata` to reduce "noise" ?

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -60,22 +63,83 @@ export type VeeamSOSApi = {
+    quotaMax: bigint | number,
+};
+
+export type ParsableBucketMetadata = Omit<Omit<BucketMetadata, 'quotaMax'>, 'capabilities'> & {
```
probably same problem as `CapacityInfoExported` : should find and use a single convention, and better express "where" this is used --> maybe `BucketMetadataJson` and `CapacityInfoJson`, or other suffix (`Serializable`, `Safe`, ...)

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -293,7 +379,7 @@ export default class BucketInfo {
         assert.strictEqual(areTagsValid(tags), true);
         if (quotaMax) {
-            assert.strictEqual(typeof quotaMax, 'number');
+            assert(typeof quotaMax === 'bigint' || typeof quotaMax === 'number');
```
duplicated & inconsistent check with line 343:
```
        if (quotaMax) {
            assert.strictEqual(typeof quotaMax, 'bigint', 'Quota must be a BigInt');
            assert(quotaMax >= 0, 'Quota cannot be negative');
        }
```

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -323,8 +409,17 @@ export default class BucketInfo {
+                CapacityInfo: convertedCapacityInfo,
+            }
+        } : undefined;
+    
```
nit:

```suggestion
        this._capabilities = capabilities && {
            ...capabilities,
            VeeamSOSApi: {
                ...capabilities?.VeeamSOSApi,
                CapacityInfo: convertedCapacityInfo,
            }
        };
    
```

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -359,8 +454,16 @@ export default class BucketInfo {
+                        this._capabilities.VeeamSOSApi.CapacityInfo
+                    ),
+                } : undefined,
+            } : undefined,
```
nit:
```suggestion
            capabilities: this._capabilities && {
                ...this._capabilities,
                VeeamSOSApi: this._capabilities.VeeamSOSApi && {
                    ...this._capabilities.VeeamSOSApi,
                    CapacityInfo: BucketInfo.serializeCapacityInfo(
                        this._capabilities.VeeamSOSApi.CapacityInfo
                    ),
                },
            },
```

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -377,6 +480,13 @@ export default class BucketInfo {
+                ...obj.capabilities?.VeeamSOSApi,
+                CapacityInfo: this.convertCapacityToBigInt(obj.capabilities?.VeeamSOSApi?.CapacityInfo),
+            },
+        };
```
should we "create" the fields, if not provided; or keep the "status-quo", as before?

```suggestion
        const capabilities: CapabilitiesInput = obj.capabilities && {
            ...obj.capabilities,
            VeeamSOSApi: obj.capabilities.VeeamSOSApi && {
                ...obj.capabilities.VeeamSOSApi,
                CapacityInfo: this.convertCapacityToBigInt(obj.capabilities.VeeamSOSApi.CapacityInfo),
            },
        };
```

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -405,6 +515,13 @@ export default class BucketInfo {
+                ...data._capabilities?.VeeamSOSApi,
+                CapacityInfo: this.convertCapacityToBigInt(data._capabilities?.VeeamSOSApi?.CapacityInfo),
+            },
+        };
```
same:
```suggestion
        const capabilities: CapabilitiesInput = data._capabilities && {
            ...data._capabilities,
            VeeamSOSApi: data._capabilities.VeeamSOSApi && {
                ...data._capabilities.VeeamSOSApi,
                CapacityInfo: this.convertCapacityToBigInt(data._capabilities.VeeamSOSApi.CapacityInfo),
            },
        };
```

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -414,8 +531,39 @@ export default class BucketInfo {
+     * @param capacityInfo - capacity info to convert
+     * @returns - converted capacity info
+     */
+    static convertCapacityToBigInt(capacityInfo: VeeamSOSApiSchema['CapacityInfo']): VeeamSOSApi['CapacityInfo'] | undefined {
```
* should this one not work when the values are actually formatted as `string` ?
i.e. so we can do a round-trip `convertCapacityToBigInt(serializeCapacityInfo(convertCapacityToBigInt(...)))`

* could we move these 2 functions are methods of `VeeamSOSApiSchema['CapacityInfo']` and `VeeamSOSApi['CapacityInfo']`, for better locally and readability?

* name should match the convention used for the differentiating the flavours of types (e.g. if we are `VeeamSOSApi` and `VeeamSOSApiSerializable` the method should be `toSerializable`), for better readability.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -70,25 +71,198 @@ function inc(str) {
+            Available: string,
+            Used: string,
+        },
+    },
```
CapacityInfoExported

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -221,26 +403,30 @@ class MongoClientInterface {
             $set: {
                 _id: bucketName,
                 value: newBucketMD,
+                vFormat: BUCKET_VERSIONS.v0,
```
technically, this has exactly the same result ; however reading this gives the false impression that the default is `v0` : when actually most of the time is the other way around...

--> would be more readable with either set `this.defaultBucketKeyFormat` and revert to v0 for usersBucket/pensieve/mpu, or create a function to "compute" the format (like before, but with `return`) and simply use it here...

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -221,26 +403,30 @@ class MongoClientInterface {
-            payload.$set.vFormat = BUCKET_VERSIONS.v0;
+        }
+
+        if (!m) {
```
nit: should go on line 413, right after we initialize `m`

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -252,30 +438,37 @@ class MongoClientInterface {
                 // "constants.usersBucket" and "PENSIEVE" since it has already
                 // been created
                 if (bucketName !== constants.usersBucket && bucketName !== PENSIEVE) {
+                    if (!this.db) {
```
this cannot happen :
* `this.db` is set only once on succesful connect()
* if we don't connect, we won't be able to get the connection (you added a check at the beginning)
* hence if we reach here, `this.db` is set already: no need to check...

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -252,30 +438,37 @@ class MongoClientInterface {
                                     key: { _id: 1 },
                                 };
-                                return this.adminDb.command(cmd, {}).then(() => cb()).catch(err => {
+                                if (!this.adminDb) {
```
same

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -301,7 +498,27 @@ class MongoClientInterface {
+                    Available: doc.value.capabilities.VeeamSOSApi.CapacityInfo.Available.toString(),
+                    Used: doc.value.capabilities.VeeamSOSApi.CapacityInfo.Used.toString(),
+                } : undefined;
+
```
same as earlier comment from BucketInfo : can't we use `serializeCapacityInfo` instead?

or better yet: can't we fix the `FIXME` comment above, by removing both serialize JSON.stringify and BucketInfo.deSerialize and using `BucketInfo.fromObj()` instead? (may require that `Long` is supported in `convertCapacityToBigInt`)

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -381,13 +608,21 @@ class MongoClientInterface {
+    putBucketAttributes(bucketName: string, bucketMD: BucketInfo, log: werelogs.Logger, cb: NodeCallBack<void>) {
         // FIXME: there should be a version of BucketInfo.serialize()
         // that does not JSON.stringify()
         const bucketInfo = BucketInfo.fromObj(bucketMD);
```
is this needed, e.g. is input not already of type `BucketInfo` ?

or maybe we want to keep it anyway, to ensure it works as before called with a plain object with the same fields (since we don't have typescript everywhere) ? then we may be conditional depending on type:

```suggestion
const bucketInfo = bucketMD instanceof BucketInfo ? bucketMD : BucketInfo.fromObj(bucketMD);
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -1226,20 +1577,26 @@ class MongoClientInterface {
-        } else {
-            filter.$gt = versionKey;
-            filter.$lt = lastVersionKey;
         }
```
nit:


```suggestion
        const filter = vFormat === BUCKET_VERSIONS.v0 ? {
            $gt: masterKey;
            $lt: lastVersionKey;
        } : {
            $gt: versionKey,
            $lt: lastVersionKey,
        };
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -1353,16 +1726,24 @@ class MongoClientInterface {
         // Check if there are other versions available
         this.getLatestVersion(c, objName, vFormat, log, (err, version) => {
-            if (err && !err.is.NoSuchKey) {
+            if (err instanceof ArsenalError && !err.is.NoSuchKey) {
```
not needed : can only return `errors.InternalError` or `errors.NoSuchKey``

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -1114,7 +1448,7 @@ class MongoClientInterface {
                 if (!doc || doc.value.isPHD) {
                     this.getLatestVersion(c, objName, vFormat, log, (err, value) => {
-                        if (err && err.is.NoSuchKey) {
+                        if (err instanceof ArsenalError && err.is.NoSuchKey) {
```
not needed, can only return `errors.InternalError` or `errors.NoSuchKey`

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -841,7 +1137,7 @@ class MongoClientInterface {
             upsert: true,
         }).then(() => this.getLatestVersion(c, objName, params.vFormat, log, (err, mstObjVal) => {
-            if (err && err.is.NoSuchKey) {
+            if (err instanceof ArsenalError && err.is.NoSuchKey) {
```
not needed, getLatestVersion can only return `errors.InternalError` or `erros.NoSuchKey`

(note: we could statically check this by change the callback type to accept an `ArsenalError` argument)

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -361,7 +588,7 @@ class MongoClientInterface {
             this.getObject(bucketName, objName, params, log, (err, obj) => {
                 if (err) {
-                    if (err.is.NoSuchKey) {
+                    if (err instanceof ArsenalError && err.is.NoSuchKey) {
```
not needed, `getObject` only yields arsenal errors

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -1353,16 +1726,24 @@ class MongoClientInterface {
                 return cb(err);
             }
-            if ((err && err.is.NoSuchKey) || (version.isDeleteMarker && vFormat === BUCKET_VERSIONS.v1)) {
+            if ((err instanceof ArsenalError && err.is.NoSuchKey) || (version!.isDeleteMarker && vFormat === BUCKET_VERSIONS.v1)) {
```
same

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -1378,7 +1759,7 @@ class MongoClientInterface {
                         // before the repair is done, we don't want to return an error
                         // in this case
-                        if (err.is.NoSuchKey) {
+                        if (err instanceof ArsenalError && err.is.NoSuchKey) {
```
same

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -1449,7 +1838,7 @@ class MongoClientInterface {
                     // a version as we expect this case when dealing with
                     // a versioning suspended object.
-                    if (err && err.is.NoSuchKey) {
+                    if (err instanceof ArsenalError && err.is.NoSuchKey) {
```
same

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -1847,7 +2291,7 @@ class MongoClientInterface {
                 // eslint-disable-next-line no-param-reassign
                 newParams.mainStreamParams.gte = range[0];
-                newParams.secondaryStreamParams.gte = range[1];
+                newParams.secondaryStreamParams!.gte = range[1];
```
same, but without any "trick" :

```suggestion
            if (newParams.secondaryStreamParams) {
                // eslint-disable-next-line no-param-reassign
                newParams.mainStreamParams.gte = range[0];
                newParams.secondaryStreamParams.gte = range[1];
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -1971,7 +2429,7 @@ class MongoClientInterface {
     checkHealth(implName, log, cb) {
         const resp = {};
-        if (this.client && this.client.topology && this.client.topology.isConnected()) {
+        if (this.isConnected) {
```
we used to check the topology, now we rely on the `close` event : does this not change the behavior (or cause the risk there of) ?

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2020,11 +2486,11 @@ class MongoClientInterface {
         this.writeUUIDIfNotExists(uuid, log, err => {
             if (err) {
-                if (err.is.InternalError) {
+                if (err instanceof ArsenalError && err.is.InternalError) {
```
not needed, can only be errors.KeyAlreadyExists or errors.InternalError

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2127,11 +2602,11 @@ class MongoClientInterface {
+        const bucketInfos: BucketInfo[] = [];
 
-        this.db.listCollections({ type: 'collection' }).toArray().then(collInfos =>
+        this.db?.listCollections({ type: 'collection' }).toArray().then(collInfos =>
```
if `db` is null, then we will call `<null>.toArray()`, and anyway not call the callback...
--> either early fail if !this.db (like in the other functions) or assert it is not null (`this.db!.`)

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2144,7 +2619,7 @@ class MongoClientInterface {
                 return this.getBucketAttributes(bucketName, log,
                     (err, bucketInfo) => {
-                        if (err?.is?.NoSuchBucket) {
+                        if (err instanceof ArsenalError && err?.is?.NoSuchBucket) {
```
not neeeded, can only be InternalError or NoSuchBucket

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2156,7 +2631,9 @@ class MongoClientInterface {
                         }
                         bucketCount++;
-                        bucketInfos.push(bucketInfo);
+                        if (bucketInfo) {
```
not possible, value is passed unless errors... We could try to type the callback like this:

```
interface ArsenalCallback<T> {
    (err: null, result: T): void;
    (err: ArsenalError, result: undefined): void;
};
```

...but while it may ensure the callback is used (in the function) as expected, typescript may not be smart enough to understand that, or even stop type inferrence because of the overloads.

seems like a good use for null-assertion though...
```suggestion
                        bucketInfos!.push(bucketInfo);
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2282,10 +2777,14 @@ class MongoClientInterface {
                                 isTransient, log, next);
                         },
-                    ], (err, results) => {
+                    ], (err, results: any) => {
```
```suggestion
                    ], (err, results: ObjectMDStats) => {
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2282,10 +2777,14 @@ class MongoClientInterface {
                         if (err) {
                             return done(err);
                         }
+                        if (!results) {
```
cannot happen if err is not null, would be best to type assert i think...

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2328,12 +2827,12 @@ class MongoClientInterface {
+                return this.getObject(PENSIEVE, overlayConfigId, null, log, next);
             },
-        ], (err, res) => {
+        ], (err, res: any) => {
```
```suggestion
        ], (err, res: ObjectMDData) => {
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -321,15 +538,19 @@ class MongoClientInterface {
      * @return {undefined}
      */
-    getBucketVFormat(bucketName, log, cb) {
+    getBucketVFormat(bucketName: string, log: werelogs.Logger, cb: NodeCallBack<string>) {
```
should we mark this function as yielding "a bucket format" (not a string!), i.e. either BUCKET_VERSIONS.v1 or BUCKET_VERSIONS.v0 ?

it would allow removing many type assertions `vFormat!`

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2577,29 +3090,33 @@ class MongoClientInterface {
                     if (err) {
                         // unable to find an object that matches the conditions
-                        if (err.is.NoSuchKey) {
+                        if (err instanceof ArsenalError && err.is.NoSuchKey) {
```
can only be InternalError or NoSuchKey

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2708,13 +3232,17 @@ class MongoClientInterface {
             err => {
                 if (err) {
-                    if (err.codeName === 'NamespaceNotFound') {
+                    if (err instanceof MongoServerError && err.codeName === 'NamespaceNotFound') {
```
do we need the instance test?
can we do something like `err.codeName? === 'NamespaceNotFound'`, since we don't really care about the type (only that it may have a codeName field, and in that case, if the value match)?

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2734,8 +3262,18 @@ class MongoClientInterface {
-    getBucketIndexes(bucketName, log, cb) {
-        const c = this.getCollection(bucketName);
+    getBucketIndexes(bucketName: string, log: werelogs.Logger, cb: NodeCallBack<{
+        name: any;
+        keys: {
+            key: any;
+            order: any;
+        }[];
+    }[]>) {
```
these are `string`s, not `any` ?

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2753,14 +3291,22 @@ class MongoClientInterface {
     getIndexingJobs(log, cb) {
         // list active createIndexes jobs
-        this.adminDb.command({
+        this.adminDb?.command({
```
if adminDb is null-ish, `this.adminDb?.command` will return null as well, and we will still crash and not call the callback, right?
-> either assert type (`this.adminDb!.`) or yield an error early?

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2753,14 +3291,22 @@ class MongoClientInterface {
             ],
         }).then(res => {
-            const jobs = [];
+            const jobs: {
```
maybe we should migrate `utils.js` to typescript as well, to avoid `any`...

francoisferrand commented on tests/unit/models/BucketInfo.spec.js:
```diff
@@ -292,8 +293,16 @@ Object.keys(acl).forEach(
-                    quotaMax: dummyBucket._quotaMax,
+                    capabilities: dummyBucket._capabilities ? {
+                        ...dummyBucket._capabilities,
+                        VeeamSOSApi: dummyBucket._capabilities.VeeamSOSApi ? {
+                            ...dummyBucket._capabilities.VeeamSOSApi,
+                            CapacityInfo: BucketInfo.serializeCapacityInfo(
+                                dummyBucket._capabilities.VeeamSOSApi.CapacityInfo,
+                            ),
+                        } : undefined,
```
we have this multiple times, probably should have another dedicated function `BucketInfo.serializeCapabilities()` or `Capabilities.serialize()`

francoisferrand commented on tests/unit/models/BucketInfo.spec.js:
```diff
@@ -292,8 +293,16 @@ Object.keys(acl).forEach(
                     tags: dummyBucket._tags,
-                    capabilities: dummyBucket._capabilities,
-                    quotaMax: dummyBucket._quotaMax,
+                    capabilities: dummyBucket._capabilities ? {
+                        ...dummyBucket._capabilities,
+                        VeeamSOSApi: dummyBucket._capabilities.VeeamSOSApi ? {
+                            ...dummyBucket._capabilities.VeeamSOSApi,
```
```suggestion
                    capabilities: dummyBucket._capabilities && {
                        ...dummyBucket._capabilities,
                        VeeamSOSApi: dummyBucket._capabilities.VeeamSOSApi && {
                            ...dummyBucket._capabilities.VeeamSOSApi,
```

Seragonia committed:
Improve typescript migration

- Better callback types, to statically check possible udnefined
  results.
- Capability for Veeam to become a set of static functions and
  types.
- Improving the performance of Bucket serialization and
  deserialization when getting/updating the bucket metadata

Issue: ARSN-457

williamlardier commented:
/wait
/approve

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -38,44 +39,54 @@ export type VersioningConfiguration = {
+    quotaMax: bigint | number,
+};
+
+export type BucketMetadataJSON = Omit<Omit<BucketMetadata, 'quotaMax'>, 'capabilities'> & {
```
```suggestion
export type BucketMetadataJSON = Omit<BucketMetadata, 'quotaMax' | 'capabilities'> & {
```

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -38,44 +39,54 @@ export type VersioningConfiguration = {
+export type BucketMetadataJSON = Omit<Omit<BucketMetadata, 'quotaMax'>, 'capabilities'> & {
+    quotaMax: string;
+    capabilities: {
+        VeeamSOSApi?: VeeamSOSApiSchema,
```
is this not the same as the "original" type, did you mean `VeeamSOSApiSerializable` ?

```suggestion
        VeeamSOSApi?: VeeamSOSApiSerializable,
```

williamlardier replied:
Updated and adapted the logic post type change

francoisferrand commented on lib/models/BucketInfo.ts:
```diff
@@ -102,7 +113,7 @@ export default class BucketInfo {
     _ingestion: { status: 'enabled' | 'disabled' } | null;
     _capabilities?: Capabilities;
-    _quotaMax: number | 0;
+    _quotaMax: bigint;
```
quotaMax may not be set?

```suggestion
    _quotaMax?: bigint;
```

williamlardier replied:
The quota is always set, with a default to 0 (means no quota)

francoisferrand replied:
it is true even in case of "upgrade", i.e. if the bucket was created before quotas were introduced? (and so mongo may not have the field)
i.e. is this structure what we have in memory (where indeed it may also be set), or what we have in mongo (where it may be missing)

williamlardier replied:
I mean, the constructor accepts an empty / void / undefined / null quota, but will always default to 0 in the class. So whatever the starting state: either we have that quota already, either it will be set to 0 at the next update.

So yes, this is fully backward compatible this way.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -70,25 +71,181 @@ function inc(str) {
+};
+
+interface ArsenalCallback<T> {
+    (err: null, result: T): void;
```
nit: could we add another overload with no paramter, when T is void ?

williamlardier replied:
I added one overload that allows `cb()` instead of `cb(null)`.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -116,15 +274,15 @@ class MongoClientInterface {
         this.defaultBucketKeyFormat = [BUCKET_VERSIONS.v0, BUCKET_VERSIONS.v1]
-            .includes(process.env.DEFAULT_BUCKET_KEY_FORMAT) ? process.env.DEFAULT_BUCKET_KEY_FORMAT
-            : BUCKET_VERSIONS.v1;
+            .includes(<BucketVersioningFormat>process.env.DEFAULT_BUCKET_KEY_FORMAT) ?
```
nit: semantically, this is not correct (we cannot assume the `env` is indeed one of the values), it should be the otherway: `[<string>BUCKET_VERSIONS.v0, <string>BUCKET_VERSIONS.v1].includes(process.env.DEFAULT_BUCKET_KEY_FORMAT)` ?

(probably does not make any difference in the end, but we never know...)

williamlardier replied:
Updated

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -1132,7 +1391,12 @@ class MongoClientInterface {
                 return next(null, doc.value);
             },
-        ], cb);
+        ], (err: ArsenalError | null | undefined, result?: ObjectMDData) => {
+            if (err) {
+                return cb(err);
+            }
+            return cb(null, result!);
+        });
```
can't anymore pass `cb` directly here, it fails type-check?
maybe we could/should use `cb!` ?

williamlardier replied:
Yes, the async type is not matching (`AsyncResultCallback<unknown, ArsenalError>`).

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2156,35 +2561,39 @@ class MongoClientInterface {
+                    method: '_getBucketInfos',
+                    error: err,
+                });
+                return cb(err);
```
this one may not be an ArsenalError, e.g. in case `listCollections().toArray()` fails?
though i don't understand why it would pass type-check :-/ (esp. when a type assertion is required on line 2569)

williamlardier replied:
I updated to return the arsenal error if the error is of this instance, otherwise, InternalError.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -2342,6 +2760,8 @@ class MongoClientInterface {
             }
             const isTransient =
+                // @ts-expect-error the locations type is not defined
+                // but is the historical check. Is this really needed?
```
indeed, the field is `location` 🤦

I guess either:
* it never really mattered because transient locations are not used in Zenko ;
* or this was working in "Zenko 1" with `locations` field, and at some point the field was renamed from locations (plural) to location, maybe to support seamless "upgrade" from 1 location to multiple?

we should create a ticket so we investigate this, both a product (is transient supposed to work) and eng (should we just look at the `location` field).

NOTE: looking at ObjectMD, `res.location` may also be a single, non-array field...

williamlardier replied:
To be done in ARSN-459, I will update the comment

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

ghost commented:
/create_integration_branches

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.2/bugfix/ARSN-457` with contents from `bugfix/ARSN-457`
and `development/8.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.2/bugfix/ARSN-457 origin/development/8.2
 $ git merge origin/bugfix/ARSN-457
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/bugfix/ARSN-457
```





*The following options are set:* **create_integration_branches**

Seragonia committed:
Update code post review

Issue: ARSN-457

Seragonia committed:
Bump project version

Issue: ARSN-457

ghost commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.2/bugfix/ARSN-457` with contents from `bugfix/ARSN-457`
and `development/8.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.2/bugfix/ARSN-457 origin/development/8.2
 $ git merge origin/bugfix/ARSN-457
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/bugfix/ARSN-457
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

Seragonia committed:
Be more safe against undefined quota or capabilities fields

Issue: ARSN-457

bert-e commented:
# History mismatch



Merge commit #9088efbb6a913d554023582a50c982e6ea835e94 on the integration branch
`w/8.2/bugfix/ARSN-457` is merging a branch which is neither the current
branch `bugfix/ARSN-457` nor the development branch
`development/8.2`.

It is likely due to a rebase of the branch `bugfix/ARSN-457` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

ghost commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.2/bugfix/ARSN-457` with contents from `bugfix/ARSN-457`
and `development/8.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.2/bugfix/ARSN-457 origin/development/8.2
 $ git merge origin/bugfix/ARSN-457
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/bugfix/ARSN-457
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

ghost commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue ARSN-457.

Goodbye williamlardier.
